spec/registry.md

```diff
@@ -359,6 +359,14 @@ the `:string` selector performs as described below.
 
 The `:string` function returns the string value of the resolved value of the _operand_.
 
+#### Composition
+
+When an _operand_ or an _option_ value uses a _variable_ with a _declaration_
+using an _expression_ with a `:string` _annotation_,
```

Comment:
This is awkward, and not quite complete. Consider:

```
.input {$x :string}
.local $y = {$x}
.local $z = {$y}
.match {$z}
```

Statically, `$z`'s declaration doesn't use an expression with a `:string` annotation, but dynamically, `$z` is bound to a value with a `:string` annotation.

My suggestion is to introduce an internal type system for resolved values: to start with, `ResolvedValue = String | Any` (this can be extended with more alternatives when other built-in functions are specified). Then you can say that anything annotated with `:string` is a `String`, and you can constrain what functions are allowed to observe about a `String` when passed one as an operand or an option value. I don't think there's a shortcut around this.

Reply:
One solution would be to forbid the un-annotated `.local`, as I proposed in #589.

Reply:
Updated with the language suggested in https://github.com/unicode-org/message-format-wg/pull/814#discussion_r1668948100